library/rmc/stubs/Rust/vec/rmc_vec.rs

```diff
@@ -0,0 +1,1085 @@
+// user can then choose to shrink_to_fit() if they want to play around with
+// tight bounds on the Vec capacity.
+const MIN_NON_ZERO_CAP: usize = 1024;
+
+// RmcVec implements a fine-grained abstraction of the Vector library for Rust.
+// It is aimed to provide a lot more functionality than the other two available
+// abstractions - NoBackVec and CVec. RmcVec aims to implement close-to-complete
+// compatibility with the Rust Standard Library (RSL) implementation.
+//
+// The goal of RMCVec is to implement basic operations of the Vec such as push(),
+// pop(), append(), insert() in a much simpler way than it is done in the RSL. The
+// intuition behind this idea is that with a simple trace, it would be much easier
+// for verification techniques such as bounded model checking to reason about
+// that piece of code. For that reason, we choose to directly drop down to libc
+// functions for low-level operations so that they can be directly translated
+// to CBMC primitives. That way, if CBMC performs better through some optimizations
+// RMC would too.
+//
+// We first implement RMCRawVec, an auxiliary data structure which holds a pointer
+// to allocated memory and the capacity of the allocation. This abstracts away
+// all low-level memory resizing operations from the actual Vec data structure.
+// It is also used later to implement RMCIter, an iterator for the RMCVec
+// data structure.
+//
+// We then use RMCRawVec as a member of Vec (RMCVec) which is the interface exposed
+// to the public. RMCVec aims to main close-to-complete compatibility with the
+// RSL Vec implementation.
+//
+// An important future work direction here is to abstract other relevant
+// data-structures such as Strings and HashMaps but implementing optimization
+// for slices seems super crucial. Most customer code deals with slices since
+// operations such as sort(), split(), get() which traditionally deal with linear
+// data structures are implemented on the slice. The advantage of doing it so
+// is that other data structures have to implement coercion to a slice and
+// can then get all these methods for free. For instance, this is done for Vec,
+// String, etc. Currently, we implement coercion to std::slice primitive type allowing
+// us to take benefits of that implementation directly but we could get much better
+// performance for real world code if we could develop abstractions for that
+// as well. Initial intuitions are that it might be harder since those operations
+// are typically not verification-friendly.
+//
+// Please note that this implementation has not been tested against ZSTs - Zero
+// Sized Types and might show unsound behavior.
+
+// RMCRawVec consists of a pointer to allocated memory and another variable tracking
+// the capacity of the allocation.
+struct RmcRawVec<T> {
+    ptr: *const T,
+    cap: usize,
+}
+
+impl<T> RmcRawVec<T> {
+    fn new() -> Self {
+        let elem_size = mem::size_of::<T>();
```

Comment:
Maybe warn or assert that this is not zero, if you're worried about soundness in that case?